persist/fs/index_lookup.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+
+	"github.com/m3db/m3db/digest"
+	"github.com/m3db/m3db/ts"
+
+	xmsgpack "github.com/m3db/m3db/persist/fs/msgpack"
+	"gopkg.in/vmihailenco/msgpack.v2"
+)
+
+// nearestIndexOffsetLookup provides a way of quickly determining the nearest offset of an
+// ID in the index file. It is not safe for concurrent use
+type nearestIndexOffsetLookup struct {
+	summaryIDsOffsets []xmsgpack.IndexSummaryToken
+	// bytes from file mmap'd into anonymous region
+	summariesMmap []byte
+	// reusable decoder stream
+	decoderStream  xmsgpack.DecoderStream
+	decoder        *xmsgpack.Decoder
```

Comment:
The `decoder *xmsgpack.Decoder` doesn't actually look used?

Reply:
You're right it was used in the new function but didn't need to be persisted in the struct. Removed